sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,171 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"errors"
+	"os"
+	"path"
+
+	"github.com/snapcore/snapd/sandbox/cgroup/inotify"
+)
+
+// appMonitorData contains all the data to monitor an specific Snap:
+// its name, the list of paths to monitor, and the channel to send
+// the notification when all the paths have been deleted.
+type appMonitorData struct {
+	name        string
+	cgroupPaths []string
+	npaths      int
+	channel     chan string
+}
+
+// CGroupMonitor allows to monitor several CGroups, detect when all
+// the running instances of each one have been closed, and notify
+// them separately. It should be considered a singleton, and
+// obtained using GetDefaultCGroupMonitor().
+type CGroupMonitor struct {
+	watched map[string][]*appMonitorData
+	watcher *inotify.Watcher
+	channel chan appMonitorData
+}
+
+var currentCGroupMonitor = CGroupMonitor{
+	watcher: nil,
+	channel: make(chan appMonitorData),
+	watched: make(map[string][]*appMonitorData),
+}
+
+func onFilesDeleted(filename string) {
+	basePath := path.Dir(filename)
+	entry := currentCGroupMonitor.watched[basePath]
```

Comment:
Something more descriptive than `entry`
```suggestion
	appWatchers := currentCGroupMonitor.watched[basePath]
```

Reply:
Done